salt/tests/unit/modules/test_metalk8s_kubernetes_utils.py

```diff
@@ -0,0 +1,211 @@
+            self.assertEqual(
+                tuple(result),
+                metalk8s_kubernetes_utils.get_kubeconfig(**kwargs)
+            )
+
+    @parameterized.expand([
+        ({"major": "1", "go_version": "go1.13.8"},),
+        ('Failed to get version info', True, True),
+    ])
+    def test_get_version_info(
+        self,
+        result,
+        get_code_raise=False,
+        raises=False,
+    ):
+        """
+        Tests the return of `get_version_info` function
+        """
+        kubeconfig_mock = MagicMock(
+            return_value=('my_kubeconfig.conf', 'my_context')
+        )
+        api_instance_mock = MagicMock()
+
+        get_code_mock = api_instance_mock.return_value.get_code
+        if get_code_raise:
+            get_code_mock.side_effect = ApiException('Failed to get version info')
+        get_code_mock.return_value.to_dict.return_value = result
+
+        with patch(
+            "metalk8s_kubernetes_utils.get_kubeconfig", kubeconfig_mock
+            ), patch(
+                "kubernetes.config.new_client_from_config", MagicMock()
+                ), patch("kubernetes.client.VersionApi", api_instance_mock):
+            if raises:
+                self.assertRaisesRegexp(
+                    CommandExecutionError,
+                    result,
+                    metalk8s_kubernetes_utils.get_version_info
+                )
+            else:
+                self.assertEqual(
+                    metalk8s_kubernetes_utils.get_version_info(),
+                    result
+                )
+            get_code_mock.assert_called()
+
+    @parameterized.expand([
+        (True,),
+        (False, True),
+    ])
+    def test_ping(
+        self,
+        result,
+        get_version_info_error=False,
+    ):
+        """
+        Tests the return of `ping` function
+        """
+        version_info_mock = MagicMock()
+        if get_version_info_error:
+            version_info_mock.side_effect = CommandExecutionError(
+                'Failed to get version info'
+            )
+        with patch(
+            "metalk8s_kubernetes_utils.get_version_info", version_info_mock
+        ):
+
+            self.assertEqual(
+                result,
+                metalk8s_kubernetes_utils.ping(),
+            )
+
+    @parameterized.expand([
+        param.explicit(kwargs=test_case)
+        for test_case in YAML_TESTS_CASES["read_and_render_yaml_file"]
+    ])
+    def test_read_and_render_yaml_file(
+        self,
+        source,
+        result,
+        context=None,
+        template=None,
+        raises=False,
+        **kwargs
+    ):
+        """
+        Tests the return of `read_and_render_yaml_file` function
+        """
+        cache_file_mock = MagicMock()
+        open_file_mock = mock_open(read_data=yaml.dump(source))
+
+        if source is None:
+            cache_file_mock.side_effect = CommandExecutionError(
+                'Source file my-source-file not found'
+            )
```

Comment:
???
No ! That it was you want to get when source empty, you should not raise it in the mock
Always return `source` here is enough